Fallback to posix version when (g)readlink -f is not available

ko1nksm opened the pull request:
Fallback to [posix version](https://github.com/ko1nksm/readlinkf) instead of using `realpath` when `(g)readlink -f` is not available. Use `(g) readlink -f` when it is available, which gives slightly better performance than before.

Fixes #307

- [x] I have reviewed the [Contributor Guidelines][contributor].
- [x] I have reviewed the [Code of Conduct][coc] and agree to abide by it

[contributor]: https://github.com/bats-core/bats-core/blob/master/docs/CONTRIBUTING.md
[coc]:         https://github.com/bats-core/bats-core/blob/master/docs/CODE_OF_CONDUCT.md

sublimino commented:
Thank you @ko1nksm :pray: 

> In order not to lengthen YOUR script, I write it code-golf-like style. Should I change to a readable style?

Yes please, that would be deeply appreciated. Do you know if the code-golf increases performance in your library?

> This code is originally written for a POSIX compliant shell, so it contains workarounds for other shells (ksh, etc.). Should I remove unnecessary code and dedicate it to bash only?

That's a great question. I think it will be easier to maintain if it doesn't diverge from your codebase, at the cost of some obviousness to a reader of the code who may wonder why `ksh` fixes are in `bats` :upside_down_face: 
- Do you intend to make upstream long-form (e.g. non-code-golf), or just this PR?  If this code diverges for that reason perhaps it does makes sense to remove `ksh` fixes too. 
- Does anybody else have an opinion here?

> It changes global variable p (local is not POSIX compliant). There is no problem if the function is called by "command substitution" (subshell). Should I change it to the local variable?

We try to avoid subshells for performance reasons if possible, a `local` would be best, but depends on the above questions (i.e. the level of change from your upstream repo you anticipate). If we're trying to prevent divergence we could leave this as a global possibly. 

You've turned this around so quickly, thank you for your efforts! :tada:

ko1nksm commented:
Thank you reply @sublimino

I will rewrite it non-code-golf style and remove workaround for other shells.

> Do you know if the code-golf increases performance in your library?

The code-golf is meant to shorten the code, not performance. It's neither fast nor slow.

Compatibility with `readlink -f` is often a problem in shell scripts world. I did research before implementing this, but I couldn't find any reliable code.

I want to get rid of this problem entirely in shell scripts world, so I'll add code upstream to make it widely used. I will rewrite it based the current code, but it is a new code and needs testing. Please give me a few days.

ko1nksm commented:
I refactored upstream [readlinkf](https://github.com/ko1nksm/readlinkf) and rebased this commit. This is still for POSIX shell. I will happily change it to bats coding style if you want.

Note:

- The `bats_readlinkf` is intended for use via command substitution (or subshell). This is because it to match the interface with the `(g)readlink` version and preserve the current directory and variables that are implicitly changed like `PWD` and `OLDPWD`, not only variables used.
- Export the `bats_readlinkf` function to not define the same function twice. I believe `libexec/bats-core/bats` does not run directly.
- When run `bash bats`, `$0` is "bash" not absolute path. Therefore I changed it to `${BASH_SOURCE[0]}`.

sublimino commented on bin/bats:
```diff
@@ -2,68 +2,53 @@
+fallback_to_readlinkf_posix() {
+  bats_readlinkf() {
+    [ "${1:-}" ] || return 1
+    target=$1 max_symlinks=10 CDPATH=''
```
Each declaration on its own line please

ko1nksm replied:
OK, Fixed.

sublimino commented on bin/bats:
```diff
@@ -2,68 +2,53 @@
+    [ "${1:-}" ] || return 1
+    target=$1 max_symlinks=10 CDPATH=''
+
+    [ -e "${target%/}" ] || target=${1%"${1##*[!/]}"} # trim trailing slashes
```
My instinct suggests this should be quoted, I guess there's a reason it's not?

sublimino commented on bin/bats:
```diff
@@ -2,68 +2,53 @@
+    cd -P . 2>/dev/null || return 1
+    while [ "$max_symlinks" -gt 0 ] && max_symlinks=$((max_symlinks - 1)); do
+      if [ ! "$target" = "${target%/*}" ]; then
+        cd -P "${target%/*}/" 2>/dev/null || break
```
I wonder if this should be `cd -P -- ...` in case of directory names like `-Lblah`

ko1nksm replied:
Yes! Thank you! However, in very old zsh, `cd` could not accept `--`. Another way I implemented is prepend `./` except it starts with `/`.

sublimino commented on bin/bats:
```diff
@@ -2,68 +2,53 @@
+    while [ "$max_symlinks" -gt 0 ] && max_symlinks=$((max_symlinks - 1)); do
+      if [ ! "$target" = "${target%/*}" ]; then
+        cd -P "${target%/*}/" 2>/dev/null || break
+        target=${target##*/}
```
Do these direct-to-variable assignments need quoting? We tend to quote by default out of paranoia!

sublimino commented on bin/bats:
```diff
@@ -2,68 +2,53 @@
+      #   <file mode>, <number of links>, <owner name>, <group name>,
+      #   <size>, <date and time>, <pathname of link>, <contents of link>
+      # https://pubs.opengroup.org/onlinepubs/9699919799/utilities/ls.html
+      link=$(ls -dl "$target" 2>/dev/null) || break
```
Perhaps again `ls -dl -- ...`?

ko1nksm replied:
Use `--` this time.

sublimino commented on bin/bats:
```diff
@@ -2,68 +2,53 @@
+      #   <size>, <date and time>, <pathname of link>, <contents of link>
+      # https://pubs.opengroup.org/onlinepubs/9699919799/utilities/ls.html
+      link=$(ls -dl "$target" 2>/dev/null) || break
+      target=${link#*" $target -> "}
```
Another quote query

ko1nksm commented:
**About quoting**

My coding style is "Do not quote in simple assignment". Quoting isn't necessary here. And another reason, This is workarounds for some shells.

The posh has many bugs even the latest version (0.13.2).

```sh
a='foobar' b='bar'
c=${a%"$b"*} # => c is "foo" (correct)
d="${a%"$b"*}" # => d is "foobar" (incorrect)
```

The bosh has bug expect the latest version (2020-05-11).

```sh
x="a*b"
y="a[*]b"
echo ${x#"$y"} # => "a*b" (correct)
echo "${x#"$y"}" # => empty (incorrect)
```

You can only change this PR if you wish. In that case, separate a commit that you can easily see the differences.

> do you think there's an advantage to using native readlink/greadlink

Depends on symbolic link depth and environment. There is no difference in the current expected use of bats, IMO. 

CPU Core i7 3.4Gzh, PCI express SSD.

| on Linux                             | real     | user     | sys      |
| ------------------------------------ | -------- | -------- | -------- |
| `readlink -f`, symlink depth: 40     | 0m0.002s | 0m0.000s | 0m0.002s |
| `readlinkf_posix`, symlink depth: 3  | 0m0.007s | 0m0.004s | 0m0.003s |
| `readlinkf_posix`, symlink depth: 40 | 0m0.096s | 0m0.069s | 0m0.031s |

| on WSL1                              | real     | user     | sys      |
| ------------------------------------ | -------- | -------- | -------- |
| `readlink -f`, symlink depth: 3      | 0m0.009s | 0m0.000s | 0m0.016s |
| `readlink -f`, symlink depth: 40     | 0m0.018s | 0m0.000s | 0m0.031s |
| `readlinkf_posix`, symlink depth: 3  | 0m0.037s | 0m0.004s | 0m0.016s |
| `readlinkf_posix`, symlink depth: 40 | 0m0.476s | 0m0.016s | 0m0.328s |

And, in rare cases, GNU `readlink`, Busybox `readlink` and FreeBSD `readlink` may have [some different behaviors](https://github.com/ko1nksm/readlinkf/blob/master/NOTE.md).

It's okay to use bats, but if you want full compatibility then fallback should be remove.

**Other changes**

Increased max symlinks, 10 to 40. it is a same value as [MAXSYMLINKS][MAXSYMLINKS] of linux kernel.

[MAXSYMLINKS]: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/namei.h?h=v5.6&id=7111951b8d4973bda27ff663f2cf18b663d15b48#n13

**Comments**

I added comments to the code, but I am not a native English speaker. Please correct it if you found weird my English.

**Note**

Currently the upstream CI is failing, but this seems to be a problem with the msys2 package repository.
